model/riscv_insts_zbb.sail

```diff
@@ -0,0 +1,365 @@
+/* ****************************************************************** */
+/* This file specifies the instructions in the BitManip Zbb extension */
+
+
+/* ****************************************************************** */
+union clause ast = RISCV_RORIW : (bits(5), regidx, regidx)
+
+mapping clause encdec = RISCV_RORIW(shamt, rs1, rd) if haveZbb() & sizeof(xlen) == 64
+  <-> 0b0110000 @ shamt @ rs1 @ 0b101 @ rd @ 0b0011011
+
+mapping clause assembly = RISCV_RORIW(shamt, rs1, rd)
+  <-> "roriw" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ hex_bits_5(shamt)
+
+function clause execute (RISCV_RORIW(shamt, rs1, rd)) = {
+  let rs1_val = X(rs1);
+  let result : xlenbits = EXTS((rs1_val[31..0] >> shamt) | (rs1_val[31..0] << (0x20 - EXTZ(shamt))));
+  X(rd) = result;
+  RETIRE_SUCCESS
+}
+
+/* ****************************************************************** */
+union clause ast = RISCV_RORI : (bits(6), regidx, regidx)
+
+mapping clause encdec = RISCV_RORI(shamt, rs1, rd) if haveZbb() & (sizeof(xlen) == 64 | shamt[5] == bitzero)
+  <-> 0b011000 @ shamt @ rs1 @ 0b101 @ rd @ 0b0010011
+
+mapping clause assembly = RISCV_RORI(shamt, rs1, rd)
+  <-> "rori" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ hex_bits_6(shamt)
+
+function clause execute (RISCV_RORI(shamt, rs1, rd)) = {
+  let rs1_val = X(rs1);
+  let result : xlenbits = if sizeof(xlen) == 32
+                            then (rs1_val >> shamt[4..0]) | (rs1_val << (0x20 - EXTZ(shamt[4..0])))
+                            else (rs1_val >> shamt[5..0]) | (rs1_val << (0x40 - EXTZ(shamt[5..0])));
+  X(rd) = result;
+  RETIRE_SUCCESS
+}
+
+/* ****************************************************************** */
+union clause ast = ZBB_RTYPEW : (regidx, regidx, regidx, bropw)
+
+mapping clause encdec = ZBB_RTYPEW(rs2, rs1, rd, RISCV_ROLW) if haveZbb() & sizeof(xlen) == 64
+  <-> 0b0110000 @ rs2 @ rs1 @ 0b001 @ rd @ 0b0111011
+
+mapping clause encdec = ZBB_RTYPEW(rs2, rs1, rd, RISCV_RORW) if haveZbb() & sizeof(xlen) == 64
+  <-> 0b0110000 @ rs2 @ rs1 @ 0b101 @ rd @ 0b0111011
+
+mapping zbb_rtypew_mnemonic : bropw <-> string = {
+  RISCV_ROLW <-> "rolw",
+  RISCV_RORW <-> "rorw"
+}
+
+mapping clause assembly = ZBB_RTYPEW(rs2, rs1, rd, op)
+  <-> zbb_rtypew_mnemonic(op) ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
+
+function clause execute (ZBB_RTYPEW(rs2, rs1, rd, op)) = {
+  let rs1_val = X(rs1);
+  let rs2_val = X(rs2);
+  let result : xlenbits = match op {
+    RISCV_ROLW => EXTS((rs1_val[31..0] << rs2_val[4..0]) | (rs1_val[31..0] >> (0x20 - EXTZ(rs2_val[4..0])))),
+    RISCV_RORW => EXTS((rs1_val[31..0] >> rs2_val[4..0]) | (rs1_val[31..0] << (0x20 - EXTZ(rs2_val[4..0]))))
+  };
+  X(rd) = EXTS(result);
+  RETIRE_SUCCESS
+}
```

Comment:
Why don't you factor the bit selection, and rename rs2_val to shimt ? (Again, just another suggestion !)

Reply:
No particular reason. I'll update this.